### PR TITLE
Couple model of mechanics and Cahn-Hilliard equation for diffusional creep in polycrystalline materials

### PR DESCRIPTION
**Pull request by cpritam**
### Complete each of the following items before creating a PR
- Include any relevant design information for your change
- Follow [Coding Standards](http://mooseframework.org/wiki/CodeStandards/)
- Submit one or more [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/) or modify existing test cases
- Reference or close a specific issue (refs # or closes #)
  #6612

### CODE REVIEW

**Comment by permcody** [comment]
@dschwen please review

File: modules/phase_field/src/kernels/CHSplitFlux.C [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    DerivativeMaterialInterface<Kernel>(parameters),
+    _component(getParam<unsigned int>("component")),
+    _mu_var(coupled("mu")),
+    _grad_mu(coupledGradient("mu")),
+    _mobility(getMaterialProperty<RealTensorValue>("mobility_name")),
+    _c_var(coupled("c"))
+{
+  if (isCoupled("c"))
+    _dmobility_dc = &getMaterialPropertyDerivative<RealTensorValue>("mobility_name", getVar("c", 0)->name());
+
```
Comment (dschwen) [inline modules/phase_field/src/kernels/CHSplitFlux.C]: empty line

File: modules/phase_field/src/kernels/CHSplitFlux.C [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#include "CHSplitFlux.h"
+
+template<>
+InputParameters validParams<CHSplitFlux>()
+{
+  InputParameters params = validParams<Kernel>();
+  params.addClassDescription("Computes flux as nodal variable - flux = -mobility * grad(chemical_potenstial)");
```
Comment (dschwen) [inline modules/phase_field/src/kernels/CHSplitFlux.C]: Why a "nodal" variable?

File: modules/phase_field/src/materials/GBDependentTensorBase.C [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+template<>
+InputParameters validParams<GBDependentTensorBase>()
+{
+  InputParameters params = validParams<Material>();
+  params.addRequiredCoupledVar("gb", "GB variable");
+  params.addParam<Real>("bulk_parameter", 0.0, "Parameter value of bulk material");
+  params.addParam<Real>("gb_parameter", 0.0, "Parameter value at GB");
+  params.addParam<MaterialPropertyName>("gb_normal_tensor_name", "Name of GB normal tensor variable");
+  params.addParam<MaterialPropertyName>("gb_tensor_prop_name", "Name of GB tensor property");
```
Comment (dschwen) [inline modules/phase_field/src/materials/GBDependentTensorBase.C]: This is a little confusing, in the doc string you call it a variable once and a property in the next one even though both are MaterialPropertyNames. What is the difference?

File: modules/phase_field/src/materials/PhaseNormalTensor.C [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+{
+  _normal_tensor[_qp].zero();
+}
+
+void
+PhaseNormalTensor::computeQpProperties()
+{
+  _normal_tensor[_qp].zero();
+
+  Real magnitude = _grad_u[_qp].size();
```
Comment (dschwen) [inline modules/phase_field/src/materials/PhaseNormalTensor.C]: This should be `.norm()` (and `const`)

File: modules/phase_field/src/materials/PhaseNormalTensor.C [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+void
+PhaseNormalTensor::initQpStatefulProperties()
+{
+  _normal_tensor[_qp].zero();
+}
+
+void
+PhaseNormalTensor::computeQpProperties()
+{
+  _normal_tensor[_qp].zero();
```
Comment (dschwen) [inline modules/phase_field/src/materials/PhaseNormalTensor.C]: I would move this to the else case for the if below

File: modules/tensor_mechanics/include/auxkernels/RealTensorAux.h [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#ifndef REALTENSORAUX_H
+#define REALTENSORAUX_H
+
+#include "AuxKernel.h"
+
+class RealTensorAux;
```
Comment (dschwen) [inline modules/tensor_mechanics/include/auxkernels/RealTensorAux.h]: I think this is the same as `MaterialRealTensorValueAux` in framework

File: modules/tensor_mechanics/src/auxkernels/RealTensorAux.C [hunk 7]
```diff
@@ -0,0 +1,33 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#include "RealTensorAux.h"
+
+template<>
+InputParameters validParams<RealTensorAux>()
```
Comment (dschwen) [inline modules/tensor_mechanics/src/auxkernels/RealTensorAux.C]: yeah.. delete this class

File: modules/tensor_mechanics/src/materials/StressBasedChemicalPotential.C [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    (*_dchemical_potential)[_qp] = 0.0;
+}
+
+void
+StressBasedChemicalPotential::computeQpProperties()
+{
+  RankTwoTensor direction_tensor_rank_two = _direction_tensor[_qp];
+  _chemical_potential[_qp] = -_stress_old[_qp].doubleContraction(direction_tensor_rank_two) * _prefactor[_qp];
+
+  if (isParamValid("c"))
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/StressBasedChemicalPotential.C]: please cache the value of `isParamValid("c")` in a `bool` member (`_has_coupled_c`). Calls to isParamValid are expensive.

File: modules/tensor_mechanics/src/materials/SumTensorIncrements.C [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+{
+  InputParameters params = validParams<Material>();
+  params.addClassDescription("Compute tensor property by summing tensor increments");
+  params.addRequiredParam<MaterialPropertyName>("tensor_name", "Name of strain property");
+  params.addParam<std::vector<MaterialPropertyName> >("coupled_tensor_increment_names", "Name of strain increment properties");
+  return params;
+}
+
+SumTensorIncrements::SumTensorIncrements(const InputParameters & parameters) :
+  DerivativeMaterialInterface<Material>(parameters),
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/SumTensorIncrements.C]: 4 space indent for initializer lists

File: modules/phase_field/src/kernels/CHSplitFlux.C (hunk 2 above)
Comment (cpritam) [inline modules/phase_field/src/kernels/CHSplitFlux.C]: Creep strain increment is a material property that depends on the gradient of flux. By making flux as a nodal variable, the order of element shape functions can be reduced. Also for a spatially varying mobility/diffusivity, the chain-rule on derivatives can be avoided.

File: modules/phase_field/src/materials/GBDependentTensorBase.C (hunk 3 above)
Comment (cpritam) [inline modules/phase_field/src/materials/GBDependentTensorBase.C]: There is no difference. It should have been property. Modified.

File: modules/phase_field/src/kernels/CHSplitFlux.C (hunk 1 above)
Comment (cpritam) [inline modules/phase_field/src/kernels/CHSplitFlux.C]: Modified.

File: modules/phase_field/src/materials/PhaseNormalTensor.C (hunk 5 above)
Comment (cpritam) [inline modules/phase_field/src/materials/PhaseNormalTensor.C]: Modified.

File: modules/tensor_mechanics/src/materials/SumTensorIncrements.C (hunk 9 above)
Comment (cpritam) [inline modules/tensor_mechanics/src/materials/SumTensorIncrements.C]: Modified.

File: modules/tensor_mechanics/src/materials/StressBasedChemicalPotential.C (hunk 8 above)
Comment (cpritam) [inline modules/tensor_mechanics/src/materials/StressBasedChemicalPotential.C]: Done.

**Comment by dschwen** [comment]
This looks fine, but it needs a documentation page on the wiki.

**Comment by permcody** [comment]
👍

**Pull request merged**